### DISCUSSION TITLE
[Solution found out] xUnit doesn't work with .NET MAUI when you have platform services.

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/maui#20585] [Q&A] [answered]
Hello everyone :) I would like to report issues with the integration .NET MAUI project with xUnit. I don't know why but if you want to add the xUnit project into solution target .NET 8, there are a lot of issues. I did the same configuration as for .NET 6 or .NET 7, e.g. like here https://jesseliberty.com/2022/11/21/net-maui-forget-me-not-7-unit-testing/

**Comment by davidortinau** [comment]
What are the issues/errors you're getting? fyi @Redth

**Reply by tschbc** [reply]
I've also been struggling with this, and frustratingly, I wasn't able to reproduce the issue with a fresh .NET 8 MAUI 8.0.6 project.

Although, now that I think about it... I wonder if the issue only shows up if you upgrade an existing project from .NET 7 -> .NET 8? Maybe I missed something. Is there an upgrade guide, or is changing framework versions/packages in `csproj` and upgrading workloads enough?

**Reply by ghost** [reply]
To be honest... I've errors involved with platform service, but everything is really okay with it. I have these errors because in csproj file I only added` <TargetFrameworks>net8.0;net8.0-android;net8.0-ios</TargetFrameworks>` and then project is burned. I've seen that .NET 6 or .NET 7 works correctly.

**Reply by ghost** [reply]
@jfversluis I tagged you as you requested :)

**Reply by ghost** [reply]
Or maybe I should run `dotnet workload update` because my project was upgrading from .NET 7 to .NET 8.

**Reply by ghost** [reply]
Ah... Probably I need to do it. Please let me know. I have installed only this.

<img width="568" alt="image" src="https://github.com/dotnet/maui/assets/55279627/f13a4ffc-5cbe-4772-94f8-54ddcd49e792">

**Reply by jfversluis** [reply]
Thanks for the mention! But as requested: what are the issues you're seeing? I will need some details if we're going to figure out what might be going wrong here. You have a project with an existing xUnit project, update to .NET 8 and it stops working? What are the steps you took? What are (error) messages you're seeing?

**Reply by ghost** [reply]
@jfversluis Okay, I investigated a lot of things, and what I see...
1. Yesterday I upgraded MAUI to 8.0.7 version - it works well.
2. Since last time the build pipeline doesn't work. I'm getting the same errors independing on the MAUI version. Locally build works correctly.

```
Users/runner/work/1/s/MyStandards/Platforms/iOS/Renderers/BottomTabViewRenderer.cs(1,40): error CS0234: The type or namespace name 'Compatibility' does not exist in the namespace 'Microsoft.Maui.Controls.Platform' (are you missing an assembly reference?) [/Users/runner/work/1/s/MyStandards/MyStandards.csproj::TargetFramework=net8.0-ios]
/Users/runner/work/1/s/MyStandards/Platforms/iOS/Renderers/CustomShellRenderer.cs(1,40): error CS0234: The type or namespace name 'Compatibility' does not exist in the namespace 'Microsoft.Maui.Controls.Handlers' (are you missing an assembly reference?) [/Users/runner/work/1/s/MyStandards/MyStandards.csproj::TargetFramework=net8.0-ios]
/Users/runner/work/1/s/MyStandards/Platforms/iOS/Renderers/CustomShellRenderer.cs(2,40): error CS0234: The type or namespace name 'Compatibility' does not exist in the namespace 'Microsoft.Maui.Controls.Platform' (are you missing an assembly reference?) [/Users/runner/work/1/s/MyStandards/MyStandards.csproj::TargetFramework=net8.0-ios]
/Users/runner/work/1/s/MyStandards/Platforms/iOS/AppDelegate.cs(7,28): error CS0246: The type or namespace name 'MauiUIApplicationDelegate' could not be found (are you missing a using directive or an assembly reference?) [/Users/runner/work/1/s/MyStandards/MyStandards.csproj::TargetFramework=net8.0-ios]
/Users/runner/work/1/s/MyStandards/Platforms/iOS/Renderers/BottomTabViewRenderer.cs(7,38): error CS0246: The type or namespace name 'ShellItemRenderer' could not be found (are you missing a using directive or an assembly reference?) [/Users/runner/work/1/s/MyStandards/MyStandards.csproj::TargetFramework=net8.0-ios]
/Users/runner/work/1/s/MyStandards/Platforms/iOS/Renderers/BottomTabViewRenderer.cs(9,34): error CS0246: The type or namespace name 'IShellContext' could not be found (are you missing a using directive or an assembly reference?) [/Users/runner/work/1/s/MyStandards/MyStandards.csproj::TargetFramework=net8.0-ios]
/Users/runner/work/1/s/MyStandards/Platforms/iOS/Renderers/CustomShellRenderer.cs(6,36): error CS0246: The type or namespace name 'ShellRenderer' could not be found (are you missing a using directive or an assembly reference?) [/Users/runner/work/1/s/MyStandards/MyStandards.csproj::TargetFramework=net8.0-ios]
/Users/runner/work/1/s/MyStandards/Platforms/iOS/Renderers/CustomShellRenderer.cs(8,24): error CS0246: The type or namespace name 'IShellItemRenderer' could not be found (are you missing a using directive or an assembly reference?) [/Users/runner/work/1/s/MyStandards/MyStandards.csproj::TargetFramework=net8.0-ios]
```

My friend has the same issues.

3. I want to add XUnit project into solution. Earlier MAUI project was targeting to .NET 7, but I updated it to .NET 8 successfully.

If you only changed this line ` <TargetFrameworks>net8.0-android;net8.0-ios</TargetFrameworks>` to ` <TargetFrameworks>net8.0;net8.0-android;net8.0-ios</TargetFrameworks>` MAUI project gives weird errors, e.g. 
platform services must have implementation - I did my platform service according to this link https://www.youtube.com/watch?v=fsO6pXz0umc

Of course error occurs only when you use ` <TargetFrameworks>net8.0-android;net8.0-ios</TargetFrameworks>`

Two issues: 
- csproj with  ` <TargetFrameworks>net8.0;net8.0-android;net8.0-ios</TargetFrameworks>`
- pipeline doesn't work

**Reply by ghost** [reply]
@jfversluis Pipelines are back and working correctly :) Will check again adding xUnit project.

**Reply by jfversluis** [reply]
Looks like you have been hitting #20600 and this has nothing to do with the actual testing then. Let me know!

**Reply by ghost** [reply]
@jfversluis I have still issues with I want to add xUnit project into solution. Both projects are setting target .NET 8.

**Reply by jfversluis** [reply]
I really want to help you, but without any clear details it's very hard to get to the core of the problem with this. The pipeline should work again, so that is not a problem anymore? If you create a new .NET MAUI project and add a xUnit project to it, does that work or not? If not, what are the steps that you take and what is the (error) messages that you are seeing? Can you build the app? Can you build the test project? Are you using Visual Studio on Windows, on Mac? VS Code? Command-line? What is the command you use?

**Reply by ghost** [reply]
I know the report should have more details, but I don't have any specific errors that I can put here. I will try to reproduce it with a new project as you mentioned. Pipelines work, so probably there were issues from Microsoft. I work on Rider IDE. Unfortunately, still problems with xUnit, but I will try to do it with new .NET MAUI project.

**Comment by ghost** [comment]
@jfversluis I closed the discussion because when I created a new MAUI project and added the xUnit project then it worked well, so probably the problems are from my side in the solution.

**Reply by ghost** [reply]
@jfversluis I just want to share you my issues when I only added .net8.0 into `TargetFrameworks` .
It occurs when I only add this.

<img width="1640" alt="Screenshot 2024-02-20 at 14 42 36" src="https://github.com/dotnet/maui/assets/55279627/01040110-5523-4f8c-a077-1080b30dce15">

<img width="1640" alt="Screenshot 2024-02-20 at 14 42 20" src="https://github.com/dotnet/maui/assets/55279627/cb739c0a-a079-4e0f-ab22-08ffa3cd1ccc">

<img width="1640" alt="Screenshot 2024-02-20 at 14 42 30" src="https://github.com/dotnet/maui/assets/55279627/96d7c861-f7c3-4b87-8b38-d3f933dad80c">

<img width="1640" alt="Screenshot 2024-02-20 at 14 41 57" src="https://github.com/dotnet/maui/assets/55279627/d8ee2e31-88dc-4398-9b25-88296601b2d4">

**Reply by ghost** [reply, marked as answer]
@jfversluis Hello :) Finally I resolved issue with adding xUnit into solution. The problem is when you have platform service in the project. If you want to add xUnit project then you get issues involved with platform services. Solution is here https://stackoverflow.com/questions/73962418/maui-how-to-use-partial-classes-for-platform-specific-implementations-together/73963543#73963543

**Reply by jfversluis** [reply]
Perfect, glad you figured it out!

**Reply by ghost** [reply]
Yes, but it's something to fix definitely in the next versions for sure ;)
